WIP: Lint refactoring. (don't merge yet)

dare-to-recurse opened the pull request:
(Sorry about the noise with multiple PRs for the lints, but I was seemingly rather tired last night when I submitted them, as I explicitly enabled a bunch of default lints that `cargo clippy` always checks for.)

---

In this PR I propose a clean-up of all my previously submitted lints.
* Removes enabled-by-default lints.
* Add additional lints considered useful by other projects.
* Fix some markdown-related nitpicks in doc-comments.
* Mark functions as const wherever possible.

### Items up for discussion:
* I added a default deny of unsafe code and missing safety documentation despite this project not having any unsafe code currently.
This seems to be a generally sane default, as it enforces a case-by-case approval of any unsafe code, along with mandatory safety comments.

* I've furthermore left three lints disabled with a comment explaining why, as enabling them will in turn create a bunch of documentation-related warnings.

@rockysnow7:
If you feel that the above is desirable I'll update the PR to enable the documentation-related lints, and also write a draft for the documentation itself.

We can then review those changes further before finally merging this PR.

rockysnow7 commented:
All the changes you have made look good. I think it would be good to enable documentation-related lints, it's probably worth it so we don't end up with a bunch of undocumented code.

(I have relatively little knowledge of clippy and lints so I am happy to go with your judgement on this.)

dare-to-recurse commented:
I'm sort of rethinking things at the moment @rockysnow7...

Maybe it's better to just start out with the default lints clippy always checks for?

And then at a slower pace consider adding additional hand-picked lints, assuming that they bring enough positive effects on the code quality.

(On the other hand I've recently seen a [significantly larger project](https://github.com/TraceMachina/nativelink/pull/1696) begin work on going from having pretty much no linting to activating almost all lints except those with known issues around false positives. So my current **slap everything in place at once** method is clearly doable as well... I'm just not sure I'm qualified enough to pull that off)

rockysnow7 commented:
I'm happy to do whatever you want lints-wise. I don't think any of our code will suffer for linting but again I am not confident with that so it's really down to what you're comfortable with :) Just push whatever commits you want when you're ready and I will merge them

dare-to-recurse commented:
Okay then :)

I've added some stuff beyond simple lint refactoring though, so we should probably rename this PR?

Specifically I've added a build-dependency on a crate called monk, which simplifies setting up automatic checks in git-hooks.
I've then used monk to set up a requirement for all commits to be correctly formatted using `cargo fmt` and to pass `cargo clippy`, as well as a requirements that all code pushed passes `cargo test` without failing any tests.

We should maybe add a note in the _contributing_ section of README.md about this being a requirement?

What remains is to maybe activate the three lints for missing documentation and to of course write this documentation, but the scope of this PR is quite large already so maybe it would be cleaner to do this in a separate PR?

I'll resolve the merge conflicts later in the morning (it's 2AM here when I'm writing this), assuming you haven't already done so by the time I wake up.

rockysnow7 commented:
This is great, thank you!!
